torch/fx/passes/fake_tensor_prop.py

```diff
@@ -36,13 +37,41 @@ def __init__(
         self._mode = mode
         mode.epoch += 1
         mode.reset_nt_tensor_id_counter()
+        self.seen_subgraphs: OrderedSet[str] = OrderedSet()
 
     def run_node(self, n: Node):
         from torch.fx.experimental.symbolic_shapes import (
             compute_unbacked_bindings,
             rebind_unbacked,
         )
 
+        if (
+            n.op == "call_function"
+            and n.target is torch.ops.higher_order.invoke_subgraph
+            and n.args[1] not in self.seen_subgraphs
+        ):
+            # Prevent redundant fake tensor prop for invoke_subgraphs. Note that
+            # there is also fake tensor caching for the entire subgraph. This
+            # happens the next time we call `run_node` for the same subgraph,
+            # which goes through super.run_node and caches the fake tensor prop.
+            # Therefore, we are propagating fake tensor through the subgraphs
+            # twice.
+            assert isinstance(n.args[1], str)
+            assert (
+                isinstance(n.args[0], torch.fx.Node)
+                and n.args[0].op == "get_attr"
+                and isinstance(n.args[0].target, str)
+            )
+            self.seen_subgraphs.add(n.args[1])
+            operands = n.args[2:]
+            example_inputs = []
+            for operand in operands:
+                assert isinstance(operand, torch.fx.Node) and "val" in operand.meta
+                example_inputs.append(operand.meta["val"])
+            return FakeTensorProp(
+                getattr(self.module, n.args[0].target), mode=self._mode
+            ).propagate(*example_inputs)
```

Comment:
This reminds me - inductor also has a way of doing incremental faketensor updates through `FakeTensorUpdater.incremental_update`. The incremental updater might not know to properly perform incremental updates on subgraphs as well today, which could lead to silent correctness? https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_utils.py#L153

Reply:
I checked that one. So anything which is run during post_grad pass, we will recurse.

view_to_reshape is outside of post_grad. Therefore, we need this PR.

Reply:
FakeTensorUpdater doesn't work on HOPs, we should fix that at some point